pkg/object/grpcserver/runtime.go

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package grpcserver
+
+import (
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/filters/grpcproxy"
+	"github.com/megaease/easegress/pkg/graceupdate"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/supervisor"
+	"github.com/megaease/easegress/pkg/util/limitlistener"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	stateNil     stateType = "nil"
+	stateFailed  stateType = "failed"
+	stateRunning stateType = "running"
+	stateClosed  stateType = "closed"
+
+	checkFailedTimeout = 10 * time.Second
+)
+
+var (
+	errNil = fmt.Errorf("")
+	gnet   = graceupdate.Global
+)
+
+type (
+	stateType string
+
+	eventReload struct {
+		nextSuperSpec *supervisor.Spec
+		muxMapper     context.MuxMapper
+	}
+	eventCheckFailed struct{}
+	eventServeFailed struct {
+		roundNum uint64
+		err      error
+	}
+	eventClose struct{ done chan struct{} }
+
+	runtime struct {
+		superSpec *supervisor.Spec
+		s         *grpc.Server
+		spec      *Spec
+		mux       *mux
+		// status
+		state         atomic.Value // stateType
+		err           atomic.Value
+		limitListener *limitlistener.LimitListener
+		roundNum      uint64
+
+		eventChan chan interface{}
+	}
+	// Status contains all status generated by runtime, for displaying to users.
+	Status struct {
+		Health bool      `yaml:"health"`
```

Comment:
Ditto. We also use json tags in Status, which could be shown in yaml on the client side.